### PR TITLE
[ENH] Efficient out-of-bag computation per honest tree

### PR DESCRIPTION
**Pull request by adam2392**
Right now, in order to have "out-of-bag" samples in the sklearn pipeline, one needs to turn `bootstrap=True` for the forest. The `sample_weights` are adjusted accordingly to "bootstrap" the samples.

This PR will allow one to access the OOB sample indices from the forest so MIGHT can take advantage of those for estimating the population statistics (e.g. S@S98, or MI). 

Another issue is how can we make out-of-bag samples without bootstrap? This might need to be a features in our sklearn fork as it is a universal property for any supervised forest. I think we can simply leverage the 

Changes proposed in this pull request:
- Add `oob_samples_` as a fitted property that efficiently generates the OOB indices per tree

## Before submitting

- [ ] I've read and followed all steps in the [Making a pull request](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [ ] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [ ] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

- [ ] All GitHub Actions jobs for my pull request have passed.

### CODE REVIEW

**Review by adam2392 — COMMENTED** [review]
- Changes in Cython code were to fix the building of wheels in Windows
- Changes in GH actions workflows, `examples/calibration` and `doc/` are used to fix the CI workflows
- All others are fair game to review
- Cython return types from `intp_t` to `int` was done in accordance with https://github.com/scikit-learn/scikit-learn/pull/27546

**Comment by adam2392** [comment]
@sampan501 @PSSF23 @SUKI-O @YuxinB can you give this PR a review? Lmk if there's any questions.

File: sktree/ensemble/_honest_forest.py [hunk 1]
```diff
@@ -257,6 +267,10 @@ class labels (multi-output problem).
     honest_indices_ : list of lists, shape=(n_estimators, n_honest)
         Indices of training samples used to learn leaf estimates.
 
+    oob_samples_ : list of lists, shape=(n_estimators, n_samples_bootstrap)
+        The indices of training samples that are "out-of-bag". Only used
+        if ``bootstrap=False``.
```
Comment (SUKI-O) [inline sktree/ensemble/_honest_forest.py:272]: Do you mean `True`?

Comment (adam2392) [reply sktree/ensemble/_honest_forest.py:272]: Yes good catch
